src/Microsoft.TemplateEngine.Cli/TemplateResolution/TemplateListResolver.cs

```diff
@@ -181,7 +148,78 @@ public static TemplateListResolutionResult GetTemplateResolutionResult(IReadOnly
             return new TemplateListResolutionResult(commandInput.TemplateName, commandInput.Language, coreMatchedTemplates, allTemplatesInContext);
         }
 
+        public static ListOrHelpTemplateListResolutionResult GetTemplateResolutionResultForListOrHelp(IReadOnlyList<ITemplateInfo> templateInfo, IHostSpecificDataLoader hostDataLoader, INewCommandInput commandInput, string defaultLanguage)
+        {
+            IReadOnlyCollection<ITemplateMatchInfo> coreMatchedTemplates;
+
+            //we need different set of templates for help and list
+            //for list we need to show all exact and partial names by name
+            //for help if there is an exact match by shortname or name we need to show help for that exact template and also apply default language mapping in case language is not specified
+            if (commandInput.IsListFlagSpecified)
+            {
+                coreMatchedTemplates = PerformCoreTemplateQueryForList(templateInfo, hostDataLoader, commandInput, defaultLanguage);
+            }
+            else
+            {
+                coreMatchedTemplates = PerformCoreTemplateQueryForHelp(templateInfo, hostDataLoader, commandInput, defaultLanguage);
+            }
+            return new ListOrHelpTemplateListResolutionResult(commandInput.TemplateName, coreMatchedTemplates);
+        }
+
+        public static IReadOnlyCollection<ITemplateMatchInfo> PerformCoreTemplateQueryForList(IReadOnlyList<ITemplateInfo> templateInfo, IHostSpecificDataLoader hostDataLoader, INewCommandInput commandInput, string defaultLanguage)
+        {
+            IReadOnlyList<FilterableTemplateInfo> filterableTemplateInfo = SetupFilterableTemplateInfoFromTemplateInfo(templateInfo);
+
+            IReadOnlyCollection<ITemplateMatchInfo> coreMatchedTemplates = TemplateListFilter.GetTemplateMatchInfo
+            (
+                filterableTemplateInfo,
+                TemplateListFilter.PartialMatchFilter,
+                WellKnownSearchFilters.NameFilter(commandInput.TemplateName),
+                //TODO: check if search for Classification is needed
```

Comment:
Are these `TODO`s going to be part of this PR or are they meant to be resolved in long term?

Reply:
this item most likely will be resolved in long term. We need to discuss first whether we want searching tags (classification) and if yes, implement it with separate param. @AR-May already created ticket for it https://github.com/dotnet/templating/issues/2475

Reply:
Got it, thanks!